### PR TITLE
Use pip to install leb128 python module

### PR DESCRIPTION
**Pull request by sbc100**
Rather than checking in our python dependencies we can now use pip to
install them.

See https://github.com/emscripten-core/emsdk/pull/1001.

### CODE REVIEW

**Comment by sbc100** [comment]
This is will WIP, but PTAL @kripken @dschuff @juj

**Comment by dschuff** [comment]
Are we pinning a particular version of the dependencies?

**Comment by sbc100** [comment]
> Are we pinning a particular version of the dependencies?

Yes, see the `requirements.txt` file.

**Comment by sbc100** [comment]
Corresponding emscripten-releases change: https://chromium-review.googlesource.com/c/emscripten-releases/+/3540615

File: em++ [hunk 1]
```diff
@@ -11,6 +11,9 @@
 # To make modifications to this file, edit `tools/run_python_compiler.sh` and
 # then run `tools/create_entry_points.py`
 
+ROOT=$(cd "$(dirname $0)" && pwd)
+export PYTHONPATH=$ROOT/.pip
```
Comment (juj) [inline em++:15]: What happens if user wants to use their own system installed python, along with their own system installed pip plugins? Looks like this is unconditionally overriding/hiding user's own PYTHONPATH?

[PYTHONPATH supports multiple paths like PATH](https://docs.python.org/3/using/cmdline.html#envvar-PYTHONPATH). Maybe prepending $ROOT/.pip as first entry would work here?

Comment (juj) [reply em++:15]: (also does LEB128 do anything else than that Wasm LEB128 encoding? If so, we could nuke using a separate module for that, and just write our own function.. [it's only a few lines](https://en.wikipedia.org/wiki/LEB128#Decode_unsigned_integer)?)

Comment (sbc100) [reply em++:15]: > What happens if user wants to use their own system installed python, along with their own system installed pip plugins? Looks like this is unconditionally overriding/hiding user's own PYTHONPATH?
> 
> [PYTHONPATH supports multiple paths like PATH](https://docs.python.org/3/using/cmdline.html#envvar-PYTHONPATH). Maybe prepending $ROOT/.pip as first entry would work here?

Yes, it would be cleaner to add to the existing PYTHONPATH.

I did test with using system install modules though and this still works if you do `pip install -r requirements`.   This is because system python modules are found via PYTHONHOME which is always search in *addition* to PYTHONPATH, so that only time this wouldn't work would be if a user did something like this:

```
$ pip install -r requirements --target some/other/place  <-- installing emscripten dependencies somewhere random
$ export PYTHONPATH=some/other/place
```

In general we want emscripten to be has hermetic as possible so there might even be and argument made that we always want explicitly ignore the users PYTHONPATH (note: not their PYTHONHOME).   For example, a users PYTHONPATH might have user module in it that conflict with emscripten's module names.

But I digress, and I think I agree, its more polite to extent PYTHONPATH than to clobber it :)

**Review by kripken — COMMENTED** [review]
I have some worry about us depending on two package managers, since we already depend on npm. We've had some issues there, and this would add more complexity.

Another approach might be to focus on rewriting code from python to js, which could avoid the need for this leb128 package in python.

I'm not suggesting that's a better course of action, but figured it's worth discussion.

**Comment by sbc100** [comment]
> I have some worry about us depending on two package managers, since we already depend on npm. We've had some issues there, and this would add more complexity.
> 
> Another approach might be to focus on rewriting code from python to js, which could avoid the need for this leb128 package in python.
> 
> I'm not suggesting that's a better course of action, but figured it's worth discussion.

I agree that having two package managers and two languages involved adds complexity, but its mostly complexity we already have and complexity that should not in any way effect emsdk users.   

This isn't really about leb128, but the ability to pull in more python modules if when we need them. Its one time hassle to setup pip, but once its working it opens up a lot of options for us.  For example we've had issues with `urllib` in the past and so we have some code that will look for `requests` if its available.   Once we have pip we can just depend on `requests` and use it unconditionally.

We already have been using pip for dev dependencies with some success.  

I agree that if we could completely re-write the python this change would not be needed.. but I don't want to block useful changes just because there is potential complete re-write in the future.

I would be more sympathetic to the "lets not add this complexity" argument if this change effected all our users, but emsdk users are not effected in this case (to them its just an implementation detail).

File: em++ (hunk 1 above)
Comment (sbc100) [reply em++]: Done

**Comment by dschuff** [comment]
For emsdk users, is the idea that we'd run pip at packaging time (and the result would go into the tarball) or that we'd run pip on the users's machine at emsdk install/update time?

**Comment by sbc100** [comment]
I defined to do with in a less intrusive way.  This PR is not way smaller and doesn't touch any of the launcher scripts.

**Comment by juj** [comment]
> but its mostly complexity we already have

I don't think this is true. Having leb128 embedded in the repo makes it more seamless for end users, rather than asking them to have to do an extra pip install step on the repository (if they don't use emsdk).

Currently it looks like Emscripten itself does not use `tools/webassembly.py` which is the only consumer of leb128? It is just an external tool offered for developers?

Even if we did have more pip packages come in for other purposes in the future, I don't think it has to mean that leb128 must come from pip as well. Those other packages might come in for some special uses that most people might not need, so they could still avoid needing to do a manual pip install step.

> I would be more sympathetic to the "lets not add this complexity" argument if this change effected all our users, but emsdk users are not effected in this case (to them its just an implementation detail).

Agree that it is great that emsdk install will deal with this automatically.

But for us this is actually primarily a legal detail. :( I need to take a route through the legal department whenever the npm manifest causes changes to license files, and with pip, the same would be true: I'll need to also audit the pip registry every time we are looking to ship to see if something has been added/updated in there we need to get approval to ship.

Secondarily, it will be a CDN package size detail. If leb128 is predistributed, it will have a nice easy known disk size that we can strip down to bare bones. If it comes from via pip install, it will likely have a lot of extra fluff. Well, very likely won't amount to anything at all in case of leb128, but in case of other packages this could be a thing. e.g. on npm side, google-closure-compiler size is a major pain.

> Once we have pip we can just depend on requests and use it unconditionally.

This is a situation I'll need to be mindful of. If we add an unconditional dependency to Emscripten to pip `requests` package, that will trigger work on my end. That is certainly not an Emscripten headache, but we have this kind of license gathering step where we need to get all the licenses from sub-components of a tool, and those get deployed into a standard structure inside a package, so that the legal team can take a package in our CDN and audit what kind of licenses/third-party components it is comprised of.

So for any new/upcoming pip dependency, I either have to implement such mechanism, then loop in a legal review, or then delete the dependency to not be packaged as part of our Emscripten redistributable.

I don't write this to mean to object, but just to illustrate how migrating a static repo-bundled leb128 to a live pip installed one generates busy work on my end. And e.g. a new `requests` package will generate more evaluation on "do we need this, can we nuke it, do we actively track its license" side, etc.

I suppose my point is for the part `Once this mechanism is in place we will be free to use whatever
pip packages we choose in emscripten's codebase.`: I hope we can actively identify (or at least remember that juj will need to :) ) the way that we use each dependency:
 - if it's a end user dependency or a Emscripten dev dependency that we can skip shipping on end users machines? and
 - even if it is a end user dependency, is it mandatory or somehow optional that if one chooses not to redistribute it, will there be a well-defined subset of Emscripten that won't be available to use? (that one might not be using - the way we currently handle Closure and its optionality is a good example)

Being clear on those can reduce my headache with legal when figuring out what to package and so on. You can appreciate how I'll probably spend an afternoon writing leb128 encoding and decoding rather than take on a legal review. :)

In summary, lgtm, though cautious about adding new non-dev-dependencies from pip.

File: tools/utils.py [hunk 2]
```diff
@@ -93,3 +93,16 @@ def write_binary(file_path, contents):
...
+# Emsdk using `emscripten/.pip` when installing these packages so add this to the
+# system path first.
+pip_path = os.path.join(path_from_root(), '.pip')
+if os.path.exists(pip_path):
+  sys.path.append(pip_path)
+
+try:
+  import leb128  # noqa
+except ImportError:
+  exit_with_error('unable to import `leb128`.  Did you run `pip install -r requirements.txt`?"')
```
Comment (juj) [inline tools/utils.py:108]: It looks like this is checking if pip is installed, even if nothing is really used from a pip package?

I think it would be good to not do this, but instead do it at actual sites where such a package is attempted to be used (i.e. like we handle closure compiler).

That way people don't need to install pip when they don't really need anything from there. Currently for example looks like tools/webassembly.py is the only consumer of leb128, and nothing uses tools/webassembly.py itself, it is just a tool provided for end users?

So this would be asking people to unconditionally do pip install even though they don't need anything from there?

**Review by juj — APPROVED** [review]

File: tools/webassembly.py [hunk 3]
```diff
@@ -8,15 +8,12 @@
 
 from collections import namedtuple
 from enum import IntEnum
+import leb128
```
[the remark this answers is not in the document]
Comment (sbc100) [reply tools/webassembly.py:11]: Sure I could do that.  The idea was to the check as early as possible, in the most widely imported file. .. that fact that the check happens use leb128 is just incidental.. it should use any of the packages listed in the requirements file for the test.

File: tools/utils.py (hunk 2 above)
Comment (sbc100) [reply tools/utils.py:108]: The webassembly.py is used in several places now, and most recently to extra the metadata from the wasm file (eliminating the dependency on wasm-emscripten-finalize) : https://github.com/emscripten-core/emscripten/pull/16529.   Its not a standalone tool but a core part of the compiler .. not optional.

**Comment by sbc100** [comment]
Thanks for the feedback @juj.   I think I understand and agree with most/all of your concerns.   Perhaps I should hold off landing this until we have a more compelling use case other than just `leb128` and `requests`?  (I seem to remember there were other use cases in on the horizon but they are escaping me now).

= Regarding the specifics of the usage of leb128 code

Its used by webassembly.py which is used for some core features of the compiler these days.

= Regarding licence checking and added headaches for you and your team

That makes sense.  I would point out that the license checking requirements are still there regardless of whether we check-in a copy of module A or pull in module A via pip/npm.   But I can see that auditing check-in code is a lot easies.

= Regarding optional dependencies

I think closure-compiler is likely to be the exception rather than rule in that the toolchain can (in theory) run without it (although I don't think we handle that complete absence of closure compile well today).   I will certainly loop you in, and consider licencing issues for any/all new user-dependencies we add.  My hope is that, in general, or in most cases, we can avoid making them optional in order to avoid having two ways to do everything.

**Comment by juj** [comment]
> Perhaps I should hold off landing this until we have a more compelling use case other than just leb128 and requests?

Thanks - that would be a great simplification for us atm.

> I would point out that the license checking requirements are still there regardless of whether we check-in a copy of module A or pull in module A

Oh yes, that goes without saying.

For us, switching from a checked-in copy of the module to a pip installed copy of the same module will likely trigger a new round trip with legal. (if any bytes in the license file changed, given the way our licensing checker for our CDN works)

Currently in our packager, we have a static code path that looks for the leb128 license file in a specific location, and copies it to our own `LICENSES/` subdirectory. (and same for other licenses) That has been easier to do statically with a static check-in of leb128. With a dynamic pip install step, we might need to make that licenses deployment mechanism more dynamic as well. That's something we can/will manage when it comes to it..

> My hope is that, in general, or in most cases, we can avoid making them optional in order to avoid having two ways to do everything.

Yeah, certainly don't want to bifurcate to two ways for achieving tasks, but rather, if there's a _sensible_ option of not using some feature that depends on a third party, then we could cater to support a setup that does not ship that third party code (it could be due to wanting to keep the distribution/download size smaller, or due to licensing impedance).
